### ISSUE TITLE
Video freezes due to constant RTP Timestamp value (from GStreamer examples)

### ISSUE THREAD

**Issue by silently** [pion/webrtc#1735]
Repo showcasing the issue: https://github.com/silently/gstreamer-transform
(it comes with a Dockerfile to help with GStreamer installation)

The intended behavior is (mainly adapted from pion examples):
stream captured from browser -> pion -> media processing through a GStreamer pipeline -> pion -> transformed stream sent and displayed in the browser

This has been tested and works for audio, currently the repo implements a unique video track. The video freezes most of the time (for periods of ~10 seconds).

A tcpdump+wireshark analysis shows most RTP packets include a constant timestamp (while Seq increases):
```
3630	11.239711	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=2668, Time=3719486720
3631	11.239739	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=2669, Time=3719486720
3632	11.239753	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=2669, Time=3719486720
3633	11.239774	192.168.0.37	192.168.0.37	RTP	1051	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=2670, Time=3719486720, Mark
3634	11.239787	192.168.0.37	192.168.0.37	RTP	1051	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=2670, Time=3719486720, Mark
```

And later:
```
10297	24.391108	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=4350, Time=3719486720
10298	24.391116	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=4350, Time=3719486720
10299	24.391148	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=4351, Time=3719486720
10300	24.391156	192.168.0.37	192.168.0.37	RTP	1242	PT=DynamicRTP-Type-96, SSRC=0x28FF75E3, Seq=4351, Time=3719486720
```
### Your environment.
  - Version: pion/webrtc v3.0.19
  - Browser: Chrome 89 and Firefox 87

It seems to me that WriteSample (on a TrackLocalStaticSample) may be responsible for this (but that does not occur with audio):
https://github.com/silently/gstreamer-transform/blob/main/gst/gst.go#L109

**Comment by silently** [comment]
After a bit of investigation, in WriteSample (around here https://github.com/pion/webrtc/blob/master/track_local_static.go#L257) it's already too late:
*  `sample.Duration.Seconds()` is -1e-09
* and `uint32(sample.Duration.Seconds() * clockRate)` is... 0

Then in the packetizer Timestamp is not updated: `p.Timestamp += samples` where samples is 0.

The problem is more likely in the gstreamer-send example around here: https://github.com/pion/example-webrtc-applications/blob/master/internal/gstreamer-src/gst.c#L54

So the issue is not related to pion/webrtc, but before closing would anyone have a recommandation about using `GST_BUFFER_DURATION(buffer)` which works with audio, but returns -1 for video/vp8.

Thanks!

**Comment by jsmouret** [comment]
not sure if that would help, but you can get the timestamp from gstreamer:
```c
GST_BUFFER_PTS(buffer),
GST_BUFFER_DURATION(buffer),
```
pass it along the duration `timestamp C.int64_t, duration C.int64_t`
and use it like:
```go
media.Sample{
	Data:      data,
	Timestamp: time.Unix(0, int64(timestamp)),
	Duration:  time.Duration(duration),
})
```

**Comment by silently** [comment]
Thanks for the hint. I tried setting Timestamp (retrieved from GST_BUFFER_PTS) but the issue remained.

I've tried something else, passing a "fake" and fixed Duration like this:
```
media.Sample{
	Data:      data,
	Timestamp: time.Unix(0, int64(timestamp)),
	Duration:  time.Duration(20000), // 20ms
})
```

And now I do receive a full (not paused) video stream in the browser (but 20ms may not be the correct value). So my guess is:

- from GStreamer point of view a video frame has no duration (GST_BUFFER_DURATION(buffer) returns -1)
- but WriteSample needs one (the solution may be to set a fix value depending on the clockRate)

Maybe it's inherent to the fact that RTP audio packets may be of variable duration, when video packets are of variable "encoding quality" ? And thus the duration is not relevant regarding video ?
